src/Storages/StorageJoin.h

```diff
@@ -31,12 +43,9 @@ class StorageJoin final : public ext::shared_ptr_helper<StorageJoin>, public Sto
     void truncate(const ASTPtr &, const StorageMetadataPtr & metadata_snapshot, const Context &, TableExclusiveLockHolder &) override;
 
     /// Access the innards.
-    HashJoinPtr & getJoin() { return join; }
+    std::shared_ptr<HashJoinHolder> getJoin() { return std::make_shared<HashJoinHolder>(rwlock, join); }
     HashJoinPtr getJoin(std::shared_ptr<TableJoin> analyzed_join) const;
```

Comment:
I think this method should actually return `HashJoinHolder` (looks a little bit more logical then having `shared_lock` inside `HashJoin`) .

Reply:
We discussed that it will requires changes in [ExpressionAnalyzer](https://github.com/ClickHouse/ClickHouse/blob/36c178c0231a073fdb3f75ca52a85ba9e4fbea7e/src/Interpreters/ExpressionAnalyzer.cpp#L808), it will have to handle this lock. So decided to keep lock in HashJoin.